### PR TITLE
fix: window.open site instance should belong to same browsing instance

### LINKED ISSUE #8100: In sandbox mode, window.opener doesn't work when children are from different domain
* Electron version: 1.4.10
* Operating system: Darwin x64

### Expected behavior
When a popup pointed to a different domain from the parent is opened. window.opener should not be null. It is required for many OAuth authentication where the parent and the child (popup) is from different domains.

### Actual behavior

When a popup is pointed towards the same domain window.opener is present but when its pointed towards a different domain window.opener is null

### How to reproduce

**main.js**
`app.on('ready', () => { console.log("ready"); const win = new BrowserWindow({webPreferences: { sandbox: true }}) win.loadURL('http://localhost:8888/index.html'); })`

**index.html**
`<a href="javascript:window.open('https://app.asana.com')">Click here</a>`

window.opener is null at app.asana.com because the domain is different from the parent

### PR DESCRIPTION
**Pull request by ppontes**
#### Description of Change
When a window is opened with window.open to navigate to a cross-site destination, the new site instance should belong to the browsing instance of the opener window in order not to break scripting relationships between windows.

fixes #8100.

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

#### Release Notes

Notes: Fixed a bug where `window.opener` of a window created with window.open from a sandboxed renderer was null.

### CODE REVIEW

File: patches/common/chromium/frame_host_manager.patch [hunk 1]
```diff
@@ -5,76 +5,65 @@ Subject: frame_host_manager.patch
...
+   scoped_refptr<SiteInstance> GetSiteInstanceForNavigation(
+       const GURL& dest_url,
 diff --git a/content/public/browser/content_browser_client.h b/content/public/browser/content_browser_client.h
 index 3be31602689c..2c22cb1cfe0d 100644
 --- a/content/public/browser/content_browser_client.h
 +++ b/content/public/browser/content_browser_client.h
 @@ -196,6 +196,15 @@ class CONTENT_EXPORT ContentBrowserClient {
   public:
    virtual ~ContentBrowserClient() {}
-
```
Comment (alexeykuzmin) [inline patches/common/chromium/frame_host_manager.patch]: Those trailing spaces matter.

**Comment by alexeykuzmin** [comment]
@ppontes Can you please check why tests fail? Those do not look like flakes.

**Comment by deepak1556** [comment]
> Adding some background on why its the current way for sandboxed renderers.

The actual site instance determination for the child web contents is made in `content::WebContentsImpl::CreateNewWindow`. The reason why its failing for cross origin is because the webcontents creation and loading happens in two different steps and the site instance is chosen incorrectly between these in electron,

* firstly after creation of the child webContents `content::WebContentsDelegate::WebContentsCreated` will be invoked

* once the setup is complete, to show the created webContents `content::WebContentsImpl::ShowCreatedWindow` will be called, which invokes `content::WebContentsDelegate::AddNewContents` and we create window with our `BrowserWindow` module.

* When `CreateGuest` in `atom/lib/browser/guest-window-manager.js` is called to load the url inside this window https://github.com/electron/electron/blob/master/lib/browser/guest-window-manager.js#L121 it will follow normal url load path and end up with a new site instance determined in `AtomBrowserClient::OverrideSiteInstanceForNavigation`

even though we short circuit site instance decision to the content layer when its a sandboxed renderer, we force creating a new one when its cross origin `AtomBrowserClient::<namespace>::IsSameWebSite`. I think the check should be removed and delegate the decision to content layer for all origins in sandboxed mode.

/cc @zcbenz

**Comment by zcbenz** [comment]
> even though we short circuit site instance decision to the content layer when its a sandboxed renderer, we force creating a new one when its cross origin `AtomBrowserClient::<namespace>::IsSameWebSite`. I think the check should be removed and delegate the decision to content layer for all origins in sandboxed mode.

Sounds  good to me.

**Comment by alexeykuzmin** [comment]
Only these two failed tests look related to the change:
1. BrowserWindow module "webPreferences" option "sandbox" option can get printer list
2. BrowserWindow module "webPreferences" option "sandbox" option should open windows in another domain with cross-scripting disabled

All others seem to be a side effect of those ^ failures.

**Comment by ppontes** [comment]
Thanks @alexeykuzmin. Checking the failing unit test.
@deepak1556, thanks for your explanation! Can you please check if I understood correctly your fix suggestion?

**Comment by deepak1556** [comment]
@ppontes the fix looks good now, thanks! once the failing specs are fixed it should be good to go.

**Comment by alexeykuzmin** [comment]
@ppontes 
I [rewrote](https://github.com/electron/electron/pull/15252) that not_very_elegant failing test. Should be easier to debug/fix.

**Comment by alexeykuzmin** [comment]
@tarruda why do we expect a `window.opener` to be `null` for a child window with a different origin?
https://github.com/electron/electron/blob/7cab7b6c7e4e0f9409372c9eea3aa6cd776d8edd/spec/api-browser-window-spec.js#L1574-L1575

**Comment by tarruda** [comment]
> @tarruda why do we expect a `window.opener` to be `null` for a child window with a different origin?

Not 100% sure, but I think it is related to the web security model which says by default pages from different origin can't share data.

**Comment by alexeykuzmin** [comment]
@ppontes Can you add a test to test precisely what is described in #8100?

**Comment by alexeykuzmin** [comment]
@ppontes If you're not going to make any more changes, please remove the `[WIP]` prefix from the PR's title.

**Comment by ppontes** [comment]
@deepak1556 can you please check the affinity fix and tests?
@alexeykuzmin updated release notes and removed wip. there seems to be a flake UT blocking the PR.

**Comment by ppontes** [comment]
@electron/docs and @electron/notifications, can you please check this PR? Thanks!

**Comment by alexeykuzmin** [comment]
@ppontes The trop bot is acting weird, can you please a manual backport to branch `4-0-x` after this PR is merged? Just a new PR targeting that branch with a cherry-pick of a merge commit of this PR.

**Comment by Gvozd** [comment]
Is it possible to backport to 3.0.x?

**Comment by alexeykuzmin** [comment]
@Gvozd We are going to. But I don't want to make any promises.

**Pull request merged**